tools/smithy-rs-tool-common/src/release_tag.rs

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use anyhow::bail;
+use lazy_static::lazy_static;
+use regex::Regex;
+use semver::Version;
+use std::cmp::Ordering;
+use std::fmt;
+use std::str::FromStr;
+
+lazy_static! {
+    static ref VERSION_TAG: Regex = Regex::new(r"^v(\d+)\.(\d+)\.(\d+)$").unwrap();
+    static ref DATE_TAG: Regex = Regex::new(r"^release-(\d{4}-\d{2}-\d{2})$").unwrap();
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct VersionReleaseTag {
+    version: Version,
+    original: String,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DateReleaseTag {
+    date: String,
+    original: String,
+}
+
+/// Represents a GitHub release tag used in aws-sdk-rust or smithy-rs.
+///
+/// Release tags can be compared with each other to see which is older/newer.
+/// Date-based tags are always considered newer than version-based tags.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ReleaseTag {
+    /// Version based tag (only used in older versions)
+    Version(VersionReleaseTag),
+    /// Date based tag
+    Date(DateReleaseTag),
+}
+
+impl ReleaseTag {
+    /// Returns the string representation of the tag
+    pub fn as_str(&self) -> &str {
+        match self {
+            ReleaseTag::Version(v) => &v.original,
+            ReleaseTag::Date(d) => &d.original,
+        }
+    }
+}
+
+impl FromStr for ReleaseTag {
+    type Err = anyhow::Error;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        if let Some(caps) = VERSION_TAG.captures(value) {
+            Ok(ReleaseTag::Version(VersionReleaseTag {
+                version: Version::new(
+                    caps.get(1).expect("validated by regex").as_str().parse()?,
+                    caps.get(2).expect("validated by regex").as_str().parse()?,
+                    caps.get(3).expect("validated by regex").as_str().parse()?,
```

Comment:
In case we ever use a release tag with a diff format, we should probably put an explanatory error here so future us have the context to address the issue.

Reply:
I think this is covered by line 72.